addons/stock/wizard/stock_assign_serial_numbers.py

```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import re
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+
+class StockAssignSerialNumbers(models.TransientModel):
+    _name = 'stock.assign.serial.numbers'
+    _description = 'Stock Assign Serial Numbers'
+
+    product_id = fields.Many2one('product.product', 'Product',
+        related='move_id.product_id', required=True)
+    move_id = fields.Many2one('stock.move', required=True)
+    next_serial_number = fields.Char('Next Serial Number')
+
+    def generate_serial_numbers(self):
+        """ This method will generate `lot_name` from a string (field
+        `next_serial_number`) and assign each `lot_name` to a move line.
+        """
+        self.ensure_one()
+        # We look if the serial number contains at least one digit.
+        caught_initial_number = re.search("\d+", self.next_serial_number)
+        if not caught_initial_number:
+            raise UserError(_('The serial number must contain at least one digit.'))
+        initial_number = caught_initial_number.group()
+        padding = len(initial_number)
+        # We split the serial number to get the prefix and suffix.
+        splitted = re.split(initial_number, self.next_serial_number)
+        prefix = splitted[0]
+        suffix = splitted[1]
+        initial_number = int(initial_number)
+
+        # Then, for each move line without `lot_id` and `lot_name`, we compute
+        # and assign the `lot_name` and set the `qty_done` to one.
+        for move_line in self.move_id.move_line_ids:
+            if not move_line.lot_id and not move_line.lot_name:
```

Comment:
it would be more elegant to read the ids of the move lines of self.stock_move_id and then generate a single write on stock move wit a command like
[(1, movelineid1, {'lot_name': x1}, (1, movelineid2, {'lot_name': x2}]